### ISSUE TITLE
imagenet workloads not converging at expected rate

### ISSUE THREAD

**Issue by priyakasimbeg** [mlcommons/algorithmic-efficiency#184]
The imagenet validation error rate vs step does not line up with reference runs for the resnet and vit workloads in both jax and pytorch.

## Description
Imagenet vit: 
- expected: 140K steps val accuracy 0.77171
- actual (jax): 187K steps val accuracy 0.5498
- actual (pytorch): 128K steps val accuracy 0.2339

Imagenet Resnet:
- expected: 140K steps val accuracy 0.77165
- actual (jax): 190K steps val accuracy 0.765680
- actual (pytorch):  pending 

## Steps to Reproduce
Run commands for Imagenet vit and Imagenet Resnet in [target setting runs](https://github.com/mlcommons/algorithmic-efficiency/blob/e6a5d0f8f4c8687aeaad0a738b9722dc8d1e5806/reference_algorithms/target_setting_algorithms/README.md).

## Source or Possible Fix
One possible source is difference in hparams, but I checked that the hparams ([imagenet resnet](reference_algorithms/target_setting_algorithms/imagenet_resnet/tuning_search_space.json) and [imagenet vit](reference_algorithms/target_setting_algorithms/imagenet_vit/tuning_search_space.json)) match the [reference hparams](https://docs.google.com/spreadsheets/d/1-OZZ_mjtevzFoNIJA5x7iJtd0egnV3BOqf5Ft1anUW8/htmlview?resourcekey=0-gj4v0Hj20h1rDILq4FZ9Gw#gid=538625624).

**Comment by znado** [comment]
We need to implement RandAug.

**Comment by pomonam** [comment]
Just for the update, the `Imagenet Resnet` workload should be ready for a re-run. We were applying `log_softmax` twice and is surprising that it could even get an accuracy of 76.5. 

In terms of the ViT, I will implement RandAug soon and then we should be ready for a re-run.

**Comment by priyakasimbeg** [comment]
Hi Juan thanks for all the hard work in resolving all the differences. I reran the Imagenet Resnet workloads and the jax workload got VERY close to the target after about 140K steps (which is what the reference internal run got and what the step hint for the learning rate schedule was set to). The target is 0.77185 and the actual max value during evals was 0.77175. We think it missed it by random chance.

![image (2)](https://user-images.githubusercontent.com/12614254/197259343-056fb57e-7a86-4e2f-8109-f4e328852b46.png)

Noticed also that the pytorch run (which ran about the same time) completed only half the number of steps I filed a separate issue for this in https://github.com/mlcommons/algorithmic-efficiency/issues/212.

So I think that in terms of fixes for Resnet we are all good, but it does beg the question whether we should relax the target value a little more. In terms of open action items for this we still want to:
1. Merge Randaug fix (@pomonam)
1. Verify that ViT Jax and Resnet are reaching the target with Randaug fix (@priyakasimbeg )
2. Verify that Pytorch does eventually reach the target, rerun for longer (@priyakasimbeg)
3. Revisit target criteria (@znado)

**Comment by pomonam** [comment]
Thanks for sharing the updates; it is good to Jax version can almost reach the target accuracy. One suggestion is that if you run the code with the flag `--profile`, it will show you a more detailed breakdown of the speed difference. 

I am still copying over imagenet dataset in my gcp environment and will be able to start soon with the merging of Randaug. Moreover, I will investigate where the speed difference is coming from for the PyTorch version. I will make the update in this thread.

**Comment by priyakasimbeg** [comment]
Reran the imagenet_vit jax and pytorch w the RandAug Fix. After about 140K steps, the Jax one gets very close (val/accuracy 0.77) but the pytorch one levels off at about 0.34. In addition the pytorch loss function looks very spikey so there is definitely still some bug in the pytorch vit workload. 

Metrics of imagenet_vit pytorch:
<img width="1057" alt="image" src="https://user-images.githubusercontent.com/12614254/197855217-c38f3962-d0ff-4586-9ac3-ec94ca557ea6.png">

**Comment by pomonam** [comment]
Thanks for sharing! I also noticed that imagenet_vit gets very close to the target accuracy, but it seems like the PyTorch one doesn't. I will look into this today.

**Comment by priyakasimbeg** [comment]
ok sounds good! Thanks again Juhan for debugging these imagenet workloads, I think we are getting close to the finish line.

**Comment by pomonam** [comment]
Priya, I believe that most issues should be fixed in #229 and it is good for a re-run. This should have also impacted PyTorch ResNet workload.

**Comment by priyakasimbeg** [comment]
Great, thank you for submitting these fixes! I will rerun the resnet jax, vit jax and vit pytorch. 

Also, yeah I remember the "loss" in tensorboard being in a weird place. Thank you for fixing that too.

**Comment by priyakasimbeg** [comment]
Update on workloads:
- Imagenet Resnet pytorch looks good after running w latest fixes, it hit the target after about 130K steps. It is still a little slow compared to the imagenet resnet jax workload (2 vs 3 steps/sec) but that will be tracked in https://github.com/mlcommons/algorithmic-efficiency/issues/212. 
- Imagenet ViT jax looks a lot better the validation accuracy gets to around 0.76 now (very close to the target ~0.77) but it did not hit it quite yet after 160K steps

So I think the remaining work on this issue now is:
- Verify Imagenet ViT pytorch workload (@priyakasimbeg)
- Investigate accuracy gap for imagenet ViT jax compared to the reference

**Comment by pomonam** [comment]
Regarding the remaining work, 
- I verified that the ImageNet ViT PyTorch workload reaches the target in 120K steps. 
- I verified that the ImageNet ViT Jax workload also reaches the target in 120K steps. I am guessing the previous gap was due to random variation, and we need to run with multiple seeds sometimes. 

I am closing this issue for now. We can discuss the slowness in a separate issue.

**Comment by priyakasimbeg** [comment]
Reopening this. I've tried to verify the imagenet resnet jax run without success.
So far 3/3 runs for me did not hit the target. 
In the last run I modified the [min aspect ratio](https://github.com/priyakasimbeg/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/imagenet_resnet/workload.py#L55) for cropping from 0.08 to 0.05 to match the internal runs but it did not hit the target either.

**Comment by pomonam** [comment]
Just to clarify, is this ResNet or ViT? I remember that ResNet for Jax was very close to the target; could you also share what accuracy you got?

**Comment by priyakasimbeg** [comment]
Sorry I meant imagenet vit

**Comment by pomonam** [comment]
Ah I see. I will send you an email with my configuration.
